src/controllers/plugins/api/JSONParse.cpp

```diff
@@ -0,0 +1,388 @@
+        return this->appendValue(ec);
+    }
+
+    bool on_null(boost::system::error_code &ec)
+    {
+        lua_pushlightuserdata(this->state, nullptr);  // json.null
+        return this->appendValue(ec);
+    }
+
+    bool on_comment_part(boost::json::string_view /* part */,
+                         boost::system::error_code & /* ec */)
+    {
+        // unhandled
+        return true;
+    }
+
+    bool on_comment(boost::json::string_view /* comment */,
+                    boost::system::error_code & /* ec */)
+    {
+        // unhandled
+        return true;
+    }
+
+private:
+    /// Append the value to the current container.
+    /// If we only got a scalar, do nothing but remember that we got some value
+    /// already.
+    bool appendValue(boost::system::error_code &ec)
+    {
+        if (this->elements.empty())
+        {
+            if (this->hadElement)
+            {
+                ec = makeCode(ParseError::MoreThanOneTopLevel);
+                return false;
+            }
+            // The element is already at the top of the stack
+            this->hadElement = true;
+            return true;
+        }
+
+        if (this->elements.back().isObject)
+        {
+            // [-3] table
+            // [-2] key
+            // [-1] value
+            lua_rawset(this->state, -3);  // table[key] = value
+        }
+        else /* array */
+        {
+            // [-2] table
+            // [-1] value
+            lua_rawseti(this->state, -2, ++this->elements.back().index);
+        }
+
+        return true;
+    }
+
+    struct Element {
+        uint16_t index : 15;  // index of the last array element
+        uint16_t isObject : 1;
+    };
+
+    lua_State *state = nullptr;
+
+    /// The stack of elements we're in.
+    ///
+    /// We need to keep track of this to issue the right call to Lua when we see
+    /// an element.
+    /// Usually this shouldn't be too deep, so allocate some inline space.
+    QVarLengthArray<Element, 16> elements;
+
+    /// Did we see any top-level element yet?
+    ///
+    /// This is only for safety. The JSON parser should guarantee that we don't
+    /// get two top-level elements.
+    bool hadElement = false;
+
+    /// A buffer for partial strings
+    std::string buffer;
+
+    friend class SaxParser;
+};
+
+class SaxParser
+{
+public:
+    SaxParser(const boost::json::parse_options &opts, lua_State *state)
+        : parser(opts, state)
+    {
+    }
+
+    void parse(std::string_view sv)
+    {
+        boost::system::error_code ec;
+        auto nRead = this->parser.write_some(false, sv.data(), sv.length(), ec);
+        if (ec)
+        {
+            luaL_error(this->parser.handler().state,
+                       "Failed to parse JSON: %s (at %s) pos: %d",
+                       ec.message().c_str(), ec.location().to_string().c_str(),
+                       static_cast<int>(nRead));
+            std::terminate();
```

Comment:
Add a comment about luaL_error's behaviour and that it behaves as if it's noreturn